File: citizensdk/android/native/src/main/kotlin/org/citizen/sdk/internal/CitizenSdkSecureStore.kt
package org.citizen.sdk.internal

import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import java.io.File
import kotlin.concurrent.withLock

/** Wallet public profile, authenticated ciphertext and permanent vault tombstones. */
internal class CitizenSdkSecureStore(directory: File) :
    CitizenSdkSqlite(directory, "secure-state-v1.sqlite3") {
    override fun createSchema(database: SQLiteDatabase) {
        database.execSQL(
            "CREATE TABLE IF NOT EXISTS wallet_profile (" +
                "wallet_index INTEGER PRIMARY KEY CHECK(wallet_index = 0), " +
                "revision INTEGER NOT NULL, record BLOB NOT NULL)",
        )
        database.execSQL(
            "CREATE TABLE IF NOT EXISTS encrypted_secret (" +
                "record_key TEXT PRIMARY KEY, revision INTEGER NOT NULL, record BLOB NOT NULL)",
        )
        database.execSQL(
            "CREATE TABLE IF NOT EXISTS vault_generation (" +
                "record_key TEXT PRIMARY KEY, wallet_index INTEGER NOT NULL, generation BLOB NOT NULL, " +
                "state INTEGER NOT NULL, operation_id BLOB NOT NULL)",
        )
    }

    fun walletProfileLoad(): CitizenSdkHostRecord = lock.withLock {
        database.query(
            "wallet_profile",
            arrayOf("revision", "record"),
            "wallet_index = 0",
            null,
            null,
            null,
            null,
        ).use { cursor ->
            if (!cursor.moveToFirst()) CitizenSdkHostRecord.absent(CitizenSdkHostDomain.WALLET_PROFILE)
            else CitizenSdkHostRecord.present(
                CitizenSdkHostDomain.WALLET_PROFILE,
                cursor.getLong(0),
                cursor.getBlob(1),
            )
        }
    }

    fun walletProfileCompareAndSwap(
        expectedRevision: Long,
        candidate: ByteArray,
    ): CitizenSdkHostRecord = transaction { db ->
        val revision = db.query(
            "wallet_profile",
            arrayOf("revision"),
            "wallet_index = 0",
            null,
            null,
            null,
            null,
        ).use { cursor -> if (cursor.moveToFirst()) cursor.getLong(0) else 0L }
        if (revision != expectedRevision || expectedRevision == Long.MAX_VALUE) {
            return@transaction CitizenSdkHostRecord.failure(CitizenSdkHostDomain.WALLET_PROFILE, 8)
        }
        val next = expectedRevision + 1L
        val values = ContentValues().apply {
            put("wallet_index", 0)
            put("revision", next)
            put("record", candidate.clone())
        }
        check(db.insertWithOnConflict("wallet_profile", null, values, SQLiteDatabase.CONFLICT_REPLACE) != -1L)
        CitizenSdkHostRecord.present(CitizenSdkHostDomain.WALLET_PROFILE, next, candidate)
    }

    fun encryptedSecretLoad(
        walletIndex: Int,
        kind: Int,
        generation: ByteArray,
        owner: ByteArray,
        accountId: ByteArray,
    ): CitizenSdkHostRecord = lock.withLock {
        val key = CitizenSdkRecordKey.secret(walletIndex, kind, generation, owner, accountId)
        database.query(
            "encrypted_secret",
            arrayOf("revision", "record"),
            "record_key = ?",
            arrayOf(key),
            null,
            null,
            null,
        ).use { cursor ->
            if (!cursor.moveToFirst()) CitizenSdkHostRecord.absent(CitizenSdkHostDomain.ENCRYPTED_SECRET_BLOB)
            else CitizenSdkHostRecord.present(
                CitizenSdkHostDomain.ENCRYPTED_SECRET_BLOB,
                cursor.getLong(0),
                cursor.getBlob(1),
            )
        }
    }

    fun encryptedSecretCompareAndSwap(
        walletIndex: Int,
        kind: Int,
        generation: ByteArray,
        owner: ByteArray,
        accountId: ByteArray,
        expectedRevision: Long,
        candidate: ByteArray,
    ): CitizenSdkHostRecord = transaction { db ->
        val key = CitizenSdkRecordKey.secret(walletIndex, kind, generation, owner, accountId)
        val revision = db.query(
            "encrypted_secret",
            arrayOf("revision"),
            "record_key = ?",
            arrayOf(key),
            null,
            null,
            null,
        ).use { cursor -> if (cursor.moveToFirst()) cursor.getLong(0) else 0L }
        if (revision != expectedRevision || expectedRevision == Long.MAX_VALUE) {
            return@transaction CitizenSdkHostRecord.failure(CitizenSdkHostDomain.ENCRYPTED_SECRET_BLOB, 8)
        }
        val next = expectedRevision + 1L
        val values = ContentValues().apply {
            put("record_key", key)
            put("revision", next)
            put("record", candidate.clone())
        }
        check(db.insertWithOnConflict("encrypted_secret", null, values, SQLiteDatabase.CONFLICT_REPLACE) != -1L)
        CitizenSdkHostRecord.present(CitizenSdkHostDomain.ENCRYPTED_SECRET_BLOB, next, candidate)
    }

    fun ensureGeneration(
        walletIndex: Int,
        generation: ByteArray,
        provisioningOperationId: ByteArray,
    ): Boolean = transaction { db ->
        val key = CitizenSdkRecordKey.walletGeneration(walletIndex, generation)
        val existing = generationRecord(db, key)
        if (existing?.first == STATE_RETIRED) return@transaction false
        if (existing?.first == STATE_ACTIVE) {
            // A generation is admitted by exactly one provisioning operation.
            // Reusing its bytes under another operation must fail closed.
            return@transaction existing.second.contentEquals(provisioningOperationId)
        }
        val values = ContentValues().apply {
            put("record_key", key)
            put("wallet_index", walletIndex)
            put("generation", generation.clone())
            put("state", STATE_ACTIVE)
            put("operation_id", provisioningOperationId.clone())
        }
        check(db.insertOrThrow("vault_generation", null, values) != -1L)
        true
    }

    fun isGenerationActive(walletIndex: Int, generation: ByteArray): Boolean = lock.withLock {
        generationState(database, CitizenSdkRecordKey.walletGeneration(walletIndex, generation)) == STATE_ACTIVE
    }

    /** Tombstone commits before the caller attempts physical Keystore deletion. */
    fun retireGeneration(
        walletIndex: Int,
        generation: ByteArray,
        cleanupOperationId: ByteArray,
    ) = transaction { db ->
        val key = CitizenSdkRecordKey.walletGeneration(walletIndex, generation)
        val values = ContentValues().apply {
            put("record_key", key)
            put("wallet_index", walletIndex)
            put("generation", generation.clone())
            put("state", STATE_RETIRED)
            put("operation_id", cleanupOperationId.clone())
        }
        check(db.insertWithOnConflict("vault_generation", null, values, SQLiteDatabase.CONFLICT_REPLACE) != -1L)
    }

    private fun generationState(db: SQLiteDatabase, key: String): Int? = db.query(
        "vault_generation",
        arrayOf("state"),
        "record_key = ?",
        arrayOf(key),
        null,
        null,
        null,
    ).use { cursor -> if (cursor.moveToFirst()) cursor.getInt(0) else null }

    private fun generationRecord(db: SQLiteDatabase, key: String): Pair<Int, ByteArray>? = db.query(
        "vault_generation",
        arrayOf("state", "operation_id"),
        "record_key = ?",
        arrayOf(key),
        null,
        null,
        null,
    ).use { cursor -> if (cursor.moveToFirst()) cursor.getInt(0) to cursor.getBlob(1) else null }

    companion object {
        internal const val STATE_ACTIVE = 1
        internal const val STATE_RETIRED = 2
    }
}

File: citizensdk/darwin/Sources/CitizenSDK/CitizenSDKSecureStore.swift
import Foundation
import SQLite3
import Darwin

/// Shares the in-process half of the vault lock between every store opened on
/// the same app-private directory. `flock` below supplies the process boundary.
private final class CitizenSDKVaultLockRegistry: @unchecked Sendable {
    private let guardLock = NSLock()
    private var locks: [String: NSLock] = [:]

    func lock(for path: String) -> NSLock {
        guardLock.lock()
        defer { guardLock.unlock() }
        if let existing = locks[path] { return existing }
        let created = NSLock()
        locks[path] = created
        return created
    }
}

/// Wallet profile, authenticated ciphertext and permanent vault tombstones.
internal final class CitizenSDKSecureStore: CitizenSDKSQLite, @unchecked Sendable {
    static let generationActive: Int64 = 1
    static let generationRetired: Int64 = 2
    private static let vaultLockRegistry = CitizenSDKVaultLockRegistry()
    private let vaultLockFile: URL
    private let vaultProcessLock: NSLock

    init(directory: URL) throws {
        vaultLockFile = directory.appendingPathComponent("vault.lock", isDirectory: false)
        vaultProcessLock = Self.vaultLockRegistry.lock(for: vaultLockFile.standardizedFileURL.path)
        try super.init(
            directory: directory,
            fileName: "secure-state-v1.sqlite3",
            schema: [
                "CREATE TABLE IF NOT EXISTS wallet_profile (wallet_index INTEGER PRIMARY KEY CHECK(wallet_index = 0), revision INTEGER NOT NULL, record BLOB NOT NULL)",
                "CREATE TABLE IF NOT EXISTS encrypted_secret (record_key TEXT PRIMARY KEY, revision INTEGER NOT NULL, record BLOB NOT NULL)",
                "CREATE TABLE IF NOT EXISTS vault_generation (record_key TEXT PRIMARY KEY, wallet_index INTEGER NOT NULL, generation BLOB NOT NULL, state INTEGER NOT NULL, operation_id BLOB NOT NULL)",
            ],
            secure: true
        )
    }

    func walletProfileLoad() throws -> CitizenSDKHostRecord {
        try read { database in
            let statement = try Self.prepare(database, "SELECT revision, record FROM wallet_profile WHERE wallet_index = 0")
            defer { sqlite3_finalize(statement) }
            if try Self.stepRowOrDone(statement) {
                return .present(.walletProfile, revision: try Self.revision(statement, 0), record: Self.data(statement, 1))
            }
            return .absent(.walletProfile)
        }
    }

    func walletProfileCAS(expected: UInt64, candidate: Data) throws -> CitizenSDKHostRecord {
        guard expected < UInt64(Int64.max) else { return .failure(.walletProfile, .conflict) }
        return try transaction { database in
            let query = try Self.prepare(database, "SELECT revision FROM wallet_profile WHERE wallet_index = 0")
            defer { sqlite3_finalize(query) }
            let actual = try Self.stepRowOrDone(query) ? try Self.revision(query, 0) : 0
            guard actual == expected else { return .failure(.walletProfile, .conflict) }
            let next = expected + 1
            let write = try Self.prepare(database, "INSERT OR REPLACE INTO wallet_profile(wallet_index, revision, record) VALUES(0, ?, ?)")
            defer { sqlite3_finalize(write) }
            try Self.bind(write, 1, Int64(next))
            try Self.bind(write, 2, candidate)
            guard sqlite3_step(write) == SQLITE_DONE else { throw CitizenSDKError(.storage, "wallet profile CAS failed") }
            return .present(.walletProfile, revision: next, record: candidate)
        }
    }

    func encryptedSecretLoad(walletIndex: UInt32, kind: UInt32, generation: Data,
                             owner: Data, accountID: Data) throws -> CitizenSDKHostRecord {
        let key = try CitizenSDKRecordKey.secret(walletIndex: walletIndex, kind: kind,
                                                 generation: generation, owner: owner, accountID: accountID)
        return try read { database in
            let statement = try Self.prepare(database, "SELECT revision, record FROM encrypted_secret WHERE record_key = ?")
            defer { sqlite3_finalize(statement) }
            try Self.bind(statement, 1, key)
            if try Self.stepRowOrDone(statement) {
                return .present(.encryptedSecretBlob, revision: try Self.revision(statement, 0), record: Self.data(statement, 1))
            }
            return .absent(.encryptedSecretBlob)
        }
    }

    func encryptedSecretCAS(walletIndex: UInt32, kind: UInt32, generation: Data,
                            owner: Data, accountID: Data, expected: UInt64,
                            candidate: Data) throws -> CitizenSDKHostRecord {
        guard expected < UInt64(Int64.max) else { return .failure(.encryptedSecretBlob, .conflict) }
        let key = try CitizenSDKRecordKey.secret(walletIndex: walletIndex, kind: kind,
                                                 generation: generation, owner: owner, accountID: accountID)
        return try transaction { database in
            let query = try Self.prepare(database, "SELECT revision FROM encrypted_secret WHERE record_key = ?")
            defer { sqlite3_finalize(query) }
            try Self.bind(query, 1, key)
            let actual = try Self.stepRowOrDone(query) ? try Self.revision(query, 0) : 0
            guard actual == expected else { return .failure(.encryptedSecretBlob, .conflict) }
            let next = expected + 1
            let write = try Self.prepare(database, "INSERT OR REPLACE INTO encrypted_secret(record_key, revision, record) VALUES(?, ?, ?)")
            defer { sqlite3_finalize(write) }
            try Self.bind(write, 1, key)
            try Self.bind(write, 2, Int64(next))
            try Self.bind(write, 3, candidate)
            guard sqlite3_step(write) == SQLITE_DONE else { throw CitizenSDKError(.storage, "encrypted secret CAS failed") }
            return .present(.encryptedSecretBlob, revision: next, record: candidate)
        }
    }

    /// A generation can be admitted by exactly one provisioning operation.
    func ensureGeneration(walletIndex: UInt32, generation: Data, operationID: Data) throws -> Bool {
        let key = try CitizenSDKRecordKey.generation(walletIndex: walletIndex, generation: generation)
        return try transaction { database in
            let query = try Self.prepare(database, "SELECT state, operation_id FROM vault_generation WHERE record_key = ?")
            defer { sqlite3_finalize(query) }
            try Self.bind(query, 1, key)
            if try Self.stepRowOrDone(query) {
                let state = sqlite3_column_int64(query, 0)
                if state == Self.generationRetired { return false }
                return state == Self.generationActive && Self.data(query, 1) == operationID
            }
            let write = try Self.prepare(database, "INSERT INTO vault_generation(record_key, wallet_index, generation, state, operation_id) VALUES(?, ?, ?, ?, ?)")
            defer { sqlite3_finalize(write) }
            try Self.bind(write, 1, key)
            try Self.bind(write, 2, Int64(walletIndex))
            try Self.bind(write, 3, generation)
            try Self.bind(write, 4, Self.generationActive)
            try Self.bind(write, 5, operationID)
            guard sqlite3_step(write) == SQLITE_DONE else { throw CitizenSDKError(.storage, "vault generation insert failed") }
            return true
        }
    }

    func isGenerationActive(walletIndex: UInt32, generation: Data) throws -> Bool {
        let key = try CitizenSDKRecordKey.generation(walletIndex: walletIndex, generation: generation)
        return try read { database in
            let query = try Self.prepare(database, "SELECT state FROM vault_generation WHERE record_key = ?")
            defer { sqlite3_finalize(query) }
            try Self.bind(query, 1, key)
            return try Self.stepRowOrDone(query) && sqlite3_column_int64(query, 0) == Self.generationActive
        }
    }

    /// Serializes the durable generation record and the corresponding physical
    /// Keychain mutation across store instances and app processes.
    func withVaultLock<T>(_ body: () throws -> T) throws -> T {
        vaultProcessLock.lock()
        defer { vaultProcessLock.unlock() }

        let descriptor = Darwin.open(vaultLockFile.path, O_CREAT | O_RDWR | O_NOFOLLOW,
                                     mode_t(S_IRUSR | S_IWUSR))
        guard descriptor >= 0 else {
            throw CitizenSDKError(.storage, "vault lock could not be opened")
        }
        defer { _ = Darwin.close(descriptor) }

        var status = stat()
        guard fstat(descriptor, &status) == 0,
              (status.st_mode & mode_t(S_IFMT)) == mode_t(S_IFREG),
              status.st_nlink == 1 else {
            throw CitizenSDKError(.storage, "vault lock is not a private regular file")
        }
        while flock(descriptor, LOCK_EX) != 0 {
            if errno == EINTR { continue }
            throw CitizenSDKError(.storage, "vault lock could not be acquired")
        }
        defer { _ = flock(descriptor, LOCK_UN) }
        return try body()
    }

    /// Tombstone is committed before physical Keychain deletion.
    func retireGeneration(walletIndex: UInt32, generation: Data, operationID: Data) throws {
        let key = try CitizenSDKRecordKey.generation(walletIndex: walletIndex, generation: generation)
        try transaction { database in
            let write = try Self.prepare(database, "INSERT OR REPLACE INTO vault_generation(record_key, wallet_index, generation, state, operation_id) VALUES(?, ?, ?, ?, ?)")
            defer { sqlite3_finalize(write) }
            try Self.bind(write, 1, key)
            try Self.bind(write, 2, Int64(walletIndex))
            try Self.bind(write, 3, generation)
            try Self.bind(write, 4, Self.generationRetired)
            try Self.bind(write, 5, operationID)
            guard sqlite3_step(write) == SQLITE_DONE else { throw CitizenSDKError(.storage, "vault generation tombstone failed") }
        }
    }
}
